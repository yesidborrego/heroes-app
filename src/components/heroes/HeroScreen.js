import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router-dom';
import { getHeroById } from '../../selectors/getHeroById';

const HeroScreen = ({ history }) => {
  const { heroId } = useParams();
  const hero = useMemo(() => getHeroById(heroId), [heroId]);

  if(!hero) return <Redirect to="/" />

  const { superhero, publisher, alter_ego, first_appearance, characters } = hero;

  const handleReturn = () => {
    (history.length <= 2) ? history.push('/') : history.goBack();
  }

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          src={`../assets/heroes/${heroId}.jpg`}
          alt={ superhero }
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>
      <div className="col-8 animate__animated animate__fadeIn">
        <h3>{ superhero }</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item"><strong>Alter ego: </strong>{ alter_ego }</li>
          <li className="list-group-item"><strong>Publisher: </strong>{ publisher }</li>
          <li className="list-group-item"><strong>First Appearance: </strong>{ first_appearance }</li>
        </ul>
        <p><strong>Characters: </strong><i>{ characters }</i></p>
        <hr />
        <button
          className="btn btn-outline-secondary"
          onClick={handleReturn}
        >
          Return
        </button>
      </div>
    </div>
  )
}

export default HeroScreen;