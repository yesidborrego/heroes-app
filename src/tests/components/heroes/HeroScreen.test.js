import '@testing-library/jest-dom';
import { mount } from "enzyme";
import { MemoryRouter, Route } from 'react-router-dom';
import HeroScreen from '../../../components/heroes/HeroScreen';


describe('Probar el componente <HeroScreen/>', () => {
  const historyMock = {
    push: jest.fn(),
    length: 10,
    goBack: jest.fn()
  };

  test('Debe mostrarse el componente "redirect" si no hay argumentos en el URL', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero']}>
        <HeroScreen history={historyMock} />
      </MemoryRouter>
    );
    expect(wrapper.find('Redirect').exists()).toBe(true);
  })

  test('Debe mostrar un heroe si el parÃ¡metro y el hero existen', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-iron']}>
        <Route path="/hero/:heroId" component={HeroScreen} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.row').exists()).toBe(true);
  })

  test('Debe regresar al inicio al hacer "click" (push)', () => {
    const historyMock = {
      push: jest.fn(),
      length: 1,
      goBack: jest.fn()
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-iron']}>
        <Route
          path="/hero/:heroId"
          component={() => <HeroScreen history={historyMock} />} />
      </MemoryRouter>
    );
    wrapper.find('button').simulate('click');
    expect(historyMock.push).toHaveBeenLastCalledWith('/');
    expect(historyMock.goBack).not.toHaveBeenCalled();
  })

  test('Debe regresar a la pantalla anterior al hacer "click" (goBack)', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-iron']}>
        <Route
          path="/hero/:heroId"
          component={() => <HeroScreen history={historyMock} />} />
      </MemoryRouter>
    );
    wrapper.find('button').simulate('click');
    expect(historyMock.push).not.toHaveBeenCalled();
    expect(historyMock.goBack).toHaveBeenCalled();
  })

  test('Debe llamar el "redirect" si el hero no existe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-iron1']}>
        <Route
          path="/hero/:heroId"
          component={() => <HeroScreen history={historyMock} />} />
      </MemoryRouter>
    );
    expect(wrapper.text()).toBe('');
  })
});
