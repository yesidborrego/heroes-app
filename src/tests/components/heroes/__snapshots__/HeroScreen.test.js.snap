// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probar el componente <HeroScreen/> Debe mostrar un heroe si el par√°metro y el hero existen 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Iron Man"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="../assets/heroes/marvel-iron.jpg"
    />
  </div>
  <div
    className="col-8 animate__animated animate__fadeIn"
  >
    <h3>
      Iron Man
    </h3>
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
        <strong>
          Alter ego: 
        </strong>
        Tony Stark
      </li>
      <li
        className="list-group-item"
      >
        <strong>
          Publisher: 
        </strong>
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <strong>
          First Appearance: 
        </strong>
        Tales of Suspense #39
      </li>
    </ul>
    <p>
      <strong>
        Characters: 
      </strong>
      <i>
        Tony Stark
      </i>
    </p>
    <hr />
    <button
      className="btn btn-outline-secondary"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;
